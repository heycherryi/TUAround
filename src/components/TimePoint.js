import React from 'react';
import PropTypes from 'prop-types';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { faCalendarDay } from '@fortawesome/free-solid-svg-icons';
import { OverlayTrigger, Tooltip } from 'react-bootstrap';
import { formatStopTimeUpdate } from '../util';
import './TimePoint.scss';

function TimePoint({ scheduleData, updateData }) {
  function formatTripTime(time) {
    const date = new Date();
    const [hour, minute, second] = time.split(':');
    date.setHours(hour);
    date.setMinutes(minute);
    date.setSeconds(second);
    return date.toLocaleTimeString([], { hour: 'numeric', minute: '2-digit' });
  }

  function renderNextDayIcon(isNextDay) {
    if (!isNextDay) {
      return;
    }
    return (
      <OverlayTrigger placement={'top'} overlay={<Tooltip>Next Day</Tooltip>}>
        <FontAwesomeIcon icon={faCalendarDay} fixedWidth={true} />
      </OverlayTrigger>
    );
  }

  // Determine which time point to use for trip update, or none.
  let updateTime = false;
  if (updateData && typeof updateData.departure !== 'undefined' && typeof updateData.departure.time === 'number') {
    updateTime = updateData.departure.time;
  } else if (updateData && typeof updateData.arrival !== 'undefined' && typeof updateData.arrival.time === 'number') {
    updateTime = updateData.arrival.time;
  }

  // Grab the relevant pieces for the scheduled time
  let scheduleTime = formatTripTime(scheduleData.departure_time);
  const isNextDay = /^2[4-9]:/.test(scheduleData.departure_time);

  // Handle case where departure and arrival time are mismatched
  let scheduleDepartNote = '';
  if (scheduleData.arrival_time !== scheduleData.departure_time) {
    scheduleTime = formatTripTime(scheduleData.arrival_time);
    scheduleDepartNote = (<> (Departs {formatTripTime(scheduleData.departure_time)})</>);
  }

  // If no update provided (e.g. past or far-future trip), return scheduled
  if (!updateTime) {
    return (
      <div className="time-point" title="Scheduled Time">
        {renderNextDayIcon(isNextDay)}
        {scheduleTime}{scheduleDepartNote}
      </div>
    );
  }
  updateTime = formatStopTimeUpdate(updateData);

  return (
    <div className="time-point">
      <div>
        <strong className="text-dark" title="Updated Time">
          {renderNextDayIcon(isNextDay)}
          {updateTime}
        </strong>
      </div>
      {scheduleTime !== updateTime && (
        <div className="small">
          <strike className="text-muted" title="Scheduled Time">
            {renderNextDayIcon(isNextDay)}
            {scheduleTime}
            {scheduleDepartNote}
          </strike>
        </div>
      )}
      {scheduleTime === updateTime && (
        <div className="small">
          (On Time)
        </div>
      )}
    </div>
  );
}

TimePoint.propTypes = {
  scheduleData: PropTypes.object.isRequired,
  updateData: PropTypes.object,
};

TimePoint.defaultProps = {
  updateData: {},
};

export default TimePoint;
