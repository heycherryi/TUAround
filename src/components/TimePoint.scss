.time-point {
  min-width: 5rem;
}
