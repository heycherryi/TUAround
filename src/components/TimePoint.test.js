/* globals test, expect */

import React from 'react';
import { render, screen } from '@testing-library/react';
import TimePoint from './TimePoint';
import tripFixture from '../fixtures/trips-270708.json';
import tripUpdateFixture from '../fixtures/trip_updates.json';

test('renders TimePoint', () => {
  const { container } = render(
    <TimePoint scheduleData={tripFixture.stop_times[0]} />,
  );
  expect(container.getElementsByClassName('text-muted')).toHaveLength(0);
  expect(screen.getByText('9:55 PM')).toBeInTheDocument();
  expect(container).toMatchSnapshot();
});

test('renders TimePoint with trip update', () => {
  const tripUpdate = tripUpdateFixture.find((i) => i.trip_update.trip.trip_id === '270708');
  const { container } = render(
    <TimePoint scheduleData={tripFixture.stop_times[0]} updateData={tripUpdate.trip_update.stop_time_update[0]} />,
  );
  expect(container.getElementsByClassName('text-muted')).toHaveLength(1);
  expect(screen.getByText('9:55 PM')).toBeInTheDocument();
  expect(screen.getByText('9:59 PM')).toBeInTheDocument();
  expect(container).toMatchSnapshot();
});
