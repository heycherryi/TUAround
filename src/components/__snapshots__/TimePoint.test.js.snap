// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders TimePoint 1`] = `
<div>
  <div
    class="time-point"
    title="Scheduled Time"
  >
    9:55 PM
  </div>
</div>
`;

exports[`renders TimePoint with trip update 1`] = `
<div>
  <div
    class="time-point"
  >
    <div>
      <strong
        class="text-dark"
        title="Updated Time"
      >
        9:59 PM
      </strong>
    </div>
    <div
      class="small"
    >
      <strike
        class="text-muted"
        title="Scheduled Time"
      >
        9:55 PM
      </strike>
    </div>
  </div>
</div>
`;
